### ISSUE TITLE
Returning '0' for Primary Key when protected $primaryKey is defined

### ISSUE THREAD

**Issue by gn0rt0n** [yajra/laravel-datatables#363]
1. Create a table such as:
  ```
Schema::create(
            'products',
            function (Blueprint $table) {
                $table->string('sku', 64)->primary('products_pkey');
                 $table->string('description');
                $table->text('body');
                $table->decimal('price', 10);
                $table->timestamp('created_at')->default(DB::RAW('CURRENT_TIMESTAMP'));
                $table->timestamp('updated_at')->default(DB::RAW('CURRENT_TIMESTAMP'));
            }
        );
  ```

2. Define in Product.php Class
  ```
 protected $primaryKey = 'sku';
  ```

3. Setup a route for the following call:
  ```
Route::get('test-products', 'ProductsController@anydata');
  ```

  and place the following function in the controller

  ```
  public function anyData()
    {
        return Datatables::of(Product::select('sku'))->make(true);
    }
  ```

4. Test the route. Results will all return '0' (I have tested using plain ::query(), as well as multiple column selection)

5. Comment out the following line in the class

  ```
protected $primaryKey = 'sku'; 
  ```

4. Test the route. Results will all be returned correctly

```
This was tested against:
* "yajra/laravel-datatables-oracle": "~6.0",
* "yajra/laravel-datatables-oracle": "dev-develop",
```

**Comment by yajra** [comment]
I just tested following your steps and it works well for me. Output below:

```json
{
draw: 0,
recordsTotal: 1,
recordsFiltered: 1,
data: [
{
sku: 123
}
],
queries: [
{
query: "select count(*) as aggregate from (select '1' as row_count from products) count_row_table",
bindings: [ ],
time: 0.56
},
{
query: "select sku from products",
bindings: [ ],
time: 0.27
}
],
input: [ ]
}
```

**Comment by yajra** [comment]
There might be something wrong with your js scripts? Can you paste your js please?

**Comment by gn0rt0n** [comment]
From my master.blade.php

```
<!-- jQuery -->
<script src="//code.jquery.com/jquery.js"></script>
<!-- DataTables -->
<script src="//cdn.datatables.net/1.10.7/js/jquery.dataTables.min.js"></script>
<!-- Bootstrap JavaScript -->
<script src="//netdna.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
<!-- App scripts -->
@stack('scripts')
```

From my index.blade.php

```
@push('scripts')
<link rel="stylesheet"
      href="https://cdn.datatables.net/buttons/1.0.3/css/buttons.dataTables.min.css">
<script src="https://cdn.datatables.net/buttons/1.0.3/js/dataTables.buttons.min.js"></script>
<script src="/vendor/datatables/buttons.server-side.js"></script>
{!! $dataTable->scripts() !!}
@stop
```

Also, not that it should matter, but did you try testing SKU as an alpha value rather than numerical?

**Comment by yajra** [comment]
It seems like you are using the service implementation? Can you please paste your `html` builder snippets? And if you are using the service approach, your index should be something like below:

```php
public function index(ProductsDataTable $dataTable)
{
    return $dataTable->render('products.index');
}
```

> Also, not that it should matter, but did you try testing SKU as an alpha value rather than numerical?

Yes, I used the same migration you posted above.

**Comment by gn0rt0n** [comment]
Here is the html builder:

```
 public function html()
    {
        return $this->builder()
            ->columns($this->getColumns())
            ->ajax('')
            ->addAction(['width' => '80px'])
            ->parameters($this->getBuilderParameters());
    }
```

I am using the service approach as you mentioned. Here is my index function:

```
public function index(ProductDataTable $dataTable)
    {
        return $dataTable->render('datatables.index');
    }
```

However, to simplify testing and just to deal with the JSON response, I created an anydata function:

```
public function anyData()
    {
        return Datatables::of(Product::select('sku'))->make(true);
    }
``` 

Which is called by:

```
Route::get('foo', 'ProductsController@anydata');
```

When I test by this method I get:

```
{
"draw": 0,
"recordsTotal": 2,
"recordsFiltered": 2,
"data": [
{
"sku": 0
},
{
"sku": 0
}
],
"queries": [
{
"query": "select count(*) as aggregate from (select '1' as \"row_count\" from \"products\" where \"products\".\"deleted_at\" is null) count_row_table",
"bindings": [],
"time": 2.65
},
{
"query": "select \"sku\" from \"products\" where \"products\".\"deleted_at\" is null",
"bindings": [],
"time": 0.9
}
],
"input": []
}
```

Additionally, I have been running in Postgres, however, I tested this against SQLite and get the same results.

Finally, thank you so much for your help on this matter. I really do appreciate it as well as the time you have put into the product overall.

**Comment by yajra** [comment]
Your code seems fine. Do you have any mutators on your product model that transforms the SKU content?

**Comment by gn0rt0n** [comment]
None. This is a very basic implementation at this point.

I also just tried it without the softdeletes enabled. The same results happened. The only difference I see between your output and mine is that the query statements are escaped. I am not sure why mine have that and yours don't. Do you think that could be related?

```
{
"draw": 0,
"recordsTotal": 6,
"recordsFiltered": 6,
"data": [
{
"sku": 0
},
{
"sku": 0
},
{
"sku": 0
},
{
"sku": 0
},
{
"sku": 0
},
{
"sku": 0
}
],
"queries": [
{
"query": "select count(*) as aggregate from (select '1' as \"row_count\" from \"products\") count_row_table",
"bindings": [],
"time": 0.65
},
{
"query": "select \"sku\" from \"products\"",
"bindings": [],
"time": 0.28
}
],
"input": []
}
```

**Comment by yajra** [comment]
I think escaping varies depending on the DB that you are using. I am using Oracle as my primary database. 

Anyways, I think the issue is connected on how Laravel handles non incrementing tables. I think you need to add `public $incrementing = false` on your model. As far as I know, Laravel casts the primary key to int by default which probably makes the value to `0`?

**Comment by gn0rt0n** [comment]
Ok, it looks like we had a breakthrough. I was testing more going between primary keys of integer (incrementing) and varchar non-incrementing, kind of along the same lines you were thinking.

Just for the sake of trying I changed the primary key from:
```
    protected $primaryKey = 'sku';
```
to this
```
    protected $primaryKey = 'public.sku';
```

and it started working. However, as I came back to report, I saw your reply, so I tried it as well - it worked too.

The following syntax is also successful:

```
protected $primaryKey = 'sku';
public $incrementing = false;
```

So, now it seems that we have two working solutions. I can't explain either, but I will dig into them more tomorrow and see what I can come up with.

Thank you again for your help.
